p2p/host/relay/autorelay.go

```diff
@@ -194,16 +192,83 @@ func (ar *AutoRelay) findRelays(ctx context.Context) {
 		}
 	}
 
+	if haveRelays == 0 {
+		// we failed to find any relays and we are not connected to any!
+		// wait a little and try again, the discovery query might have returned only dead peers
+		log.Debug("no relays connected; retrying in 30s")
+		select {
+		case <-time.After(30 * time.Second):
+			goto again
+		case <-ctx.Done():
+			return
+		}
+	}
+
 	if update > 0 || ar.addrs == nil {
 		ar.updateAddrs()
 	}
 }
 
-func (ar *AutoRelay) selectRelays(pis []pstore.PeerInfo) []pstore.PeerInfo {
+func (ar *AutoRelay) selectRelays(ctx context.Context, pis []pstore.PeerInfo, count, maxq int) []pstore.PeerInfo {
 	// TODO better relay selection strategy; this just selects random relays
 	//      but we should probably use ping latency as the selection metric
+
+	if len(pis) == 0 {
+		return pis
+	}
+
+	if len(pis) < count {
+		count = len(pis)
+	}
+
+	if len(pis) < maxq {
+		maxq = len(pis)
+	}
+
+	// only select relays that can be found by routing
+	type queryResult struct {
+		pi  pstore.PeerInfo
+		err error
+	}
+	result := make([]pstore.PeerInfo, 0, count)
+	resultCh := make(chan queryResult, maxq)
+
+	qctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
+
+	// shuffle to randomize the order of queries
 	shuffleRelays(pis)
-	return pis
+	for _, pi := range pis[:maxq] {
+		go func(p peer.ID) {
+			pi, err := ar.router.FindPeer(qctx, p)
+			if err != nil {
+				log.Debugf("error finding relay peer %s: %s", p, err.Error())
+			}
+			resultCh <- queryResult{pi: pi, err: err}
+		}(pi.ID)
+	}
+
+	rcount := 0
+	for len(result) < count && rcount < maxq {
+		select {
+		case qr := <-resultCh:
+			rcount++
+			if qr.err == nil {
+				pi := cleanupAddressSet(qr.pi)
```

Comment:
The bad addresses are already going to be in our peerstore, right?

Reply:
Maybe, but they are not going to be used in the address announcement (which is what addrsplodes te relay addr advertisement).

Reply:
that would be a very nice improvement with `RESERVE`!